File: model-client/src/main/java/org/modelix/model/util/pmap/SmallPMap.java
package org.modelix.model.util.pmap;

import org.modelix.model.util.StreamUtils;

import java.util.Collections;
import java.util.Objects;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public abstract class SmallPMap<K, V> implements CustomPMap<K, V> {
  private static final SmallPMap EMPTY = new Multiple(new Object[0], new Object[0]);
  public static <K, V> SmallPMap<K, V> empty() {
    return EMPTY;
  }
  private static <K, V> SmallPMap<K, V> create(Object[] keys, Object[] values) {
    if (keys.length == 0) {
      return empty();
    }
    if (keys.length == 1) {
      return new Single<K, V>((K) keys[0], (V) values[0]);
    }
    return new Multiple<K, V>(keys, values);
  }

  public static <K, V> SmallPMap<K, V> createS(Iterable<K> keys, Iterable<V> values) {
    return create(
            StreamUtils.toStream(keys).toArray(Object[]::new),
            StreamUtils.toStream(values).toArray(Object[]::new));
  }

  @Override
  public abstract V get(K key);
  @Override
  public abstract SmallPMap<K, V> put(K key, V value);
  @Override
  public abstract SmallPMap<K, V> remove(K key);
  @Override
  public abstract Iterable<K> keys();
  @Override
  public abstract Iterable<V> values();

  public static class Single<K, V> extends SmallPMap<K, V> {
    private K key;
    private V value;

    private Single(K key, V value) {
      this.key = key;
      this.value = value;
    }
    @Override
    public V get(K key) {
      return (Objects.equals(this.key, key) ? value : null);
    }
    @Override
    public Iterable<K> keys() {
      return Collections.singleton(key);
    }
    @Override
    public SmallPMap<K, V> put(K key, V value) {
      if (value == null) {
        return remove(key);
      }
      if (Objects.equals(key, this.key)) {
        if (Objects.equals(value, this.value)) {
          return this;
        } else {
          return new Single<K, V>(key, value);
        }
      }
      return create(new Object[]{this.key, key}, new Object[]{this.value, value});
    }
    @Override
    public SmallPMap<K, V> remove(K key) {
      return (Objects.equals(key, this.key) ? EMPTY : this);
    }
    @Override
    public Iterable<V> values() {
      return Collections.singleton(value);
    }
    @Override
    public boolean containsKey(K key) {
      return Objects.equals(key, this.key);
    }
  }

  public static class Multiple<K, V> extends SmallPMap<K, V> {
    private Object[] keys;
    private Object[] values;

    private Multiple(Object[] keys, Object[] values) {
      this.keys = keys;
      this.values = values;
    }

    @Override
    public V get(K key) {
      for (int i = 0; i < keys.length; i++) {
        if (Objects.equals(keys[i], key)) {
          return (V) values[i];
        }
      }
      return null;
    }

    @Override
    public SmallPMap<K, V> put(K key, V value) {
      if (value == null) {
        return remove(key);
      }

      int index = COWArrays.indexOf(keys, key);
      if (index != -1) {
        if (Objects.equals(value, values[index])) {
          return this;
        } else {
          return create(keys, COWArrays.set(values, index, value));
        }
      } else {
        return create(COWArrays.add(keys, key), COWArrays.add(values, value));
      }
    }

    @Override
    public SmallPMap<K, V> remove(K key) {
      int index = COWArrays.indexOf(keys, key);
      if (index != -1) {
        return create(COWArrays.removeAt(keys, index), COWArrays.removeAt(values, index));
      } else {
        return this;
      }
    }

    @Override
    public Iterable<K> keys() {
      return Stream.of(keys).map(it -> (K)it).collect(Collectors.toList());
    }

    @Override
    public Iterable<V> values() {
      return Stream.of(values).map(it -> (V)it).collect(Collectors.toList());
    }

    @Override
    public boolean containsKey(K key) {
      for (Object k : this.keys) {
        if (Objects.equals(k, key)) {
          return true;
        }
      }
      return false;
    }
  }
}

File: model-client/src/main/kotlin/org/modelix/model/util/pmap/SmallPMap.kt
package org.modelix.model.util.pmap

import org.modelix.model.util.StreamUtils
import java.util.stream.Collectors
import java.util.stream.Stream

abstract class SmallPMap<K, V> : CustomPMap<K, V> {
    abstract override fun get(key: K): V
    abstract override fun put(key: K, value: V): SmallPMap<K, V>?
    abstract override fun remove(key: K): SmallPMap<K, V>?
    abstract override fun keys(): Iterable<K>?
    abstract override fun values(): Iterable<V>?
    class Single<K, V>(private val key: K, private val value: V) : SmallPMap<K, V?>() {
        override fun get(key: K): V? {
            return if (this.key == key) value else null
        }

        override fun keys(): Iterable<K>? {
            return setOf(key)
        }

        override fun put(key: K, value: V?): SmallPMap<K, V?>? {
            if (value == null) {
                return remove(key)
            }
            return if (key == this.key) {
                if (value == this.value) {
                    this
                } else {
                    Single<K, V?>(key, value)
                }
            } else create(arrayOf(this.key, key), arrayOf(this.value, value))
        }

        override fun remove(key: K): SmallPMap<K, V?>? {
            return if (key == this.key) EMPTY else this
        }

        override fun values(): Iterable<V?>? {
            return setOf(value)
        }

        override fun containsKey(key: K): Boolean {
            return key == this.key
        }

    }

    class Multiple<K, V>(private val keys: Array<Any?>?, private val values: Array<Any?>?) : SmallPMap<K, V?>() {
        override fun get(key: K): V? {
            for (i in keys!!.indices) {
                if (keys[i] == key) {
                    return values!![i] as V?
                }
            }
            return null
        }

        override fun put(key: K, value: V?): SmallPMap<K, V?>? {
            if (value == null) {
                return remove(key)
            }
            val index = COWArrays.indexOf(keys, key)
            return if (index != -1) {
                if (value == values!![index]) {
                    this
                } else {
                    create(keys, COWArrays.set(values, index, value))
                }
            } else {
                create(COWArrays.add(keys, key), COWArrays.add(values, value))
            }
        }

        override fun remove(key: K): SmallPMap<K, V?>? {
            val index = COWArrays.indexOf(keys, key)
            return if (index != -1) {
                create(COWArrays.removeAt(keys, index), COWArrays.removeAt(values, index))
            } else {
                this
            }
        }

        override fun keys(): Iterable<K>? {
            return Stream.of(*keys).map { it: Any? -> it as K? }.collect(Collectors.toList())
        }

        override fun values(): Iterable<V?>? {
            return Stream.of(*values).map { it: Any? -> it as V? }.collect(Collectors.toList())
        }

        override fun containsKey(key: K): Boolean {
            for (k in keys!!) {
                if (k == key) {
                    return true
                }
            }
            return false
        }

    }

    companion object {
        private val EMPTY: SmallPMap<*, *> = Multiple<Any?, Any?>(arrayOfNulls(0), arrayOfNulls(0))
        fun <K, V> empty(): SmallPMap<K, V> {
            return EMPTY
        }

        private fun <K, V> create(keys: Array<Any?>?, values: Array<Any?>?): SmallPMap<K?, V?> {
            if (keys!!.size == 0) {
                return empty()
            }
            return if (keys.size == 1) {
                Single(keys[0] as K?, values!![0] as V?)
            } else Multiple<K?, V?>(keys, values)
        }

        fun <K, V> createS(keys: Iterable<K>?, values: Iterable<V>?): SmallPMap<K, V> {
            return create(
                    StreamUtils.toStream(keys).toArray { _Dummy_.__Array__() },
                    StreamUtils.toStream(values).toArray { _Dummy_.__Array__() })
        }
    }
}